File: db-commons/src/main/java/net/blueshell/db/permission/BasePermissionEvaluator.java
package net.blueshell.db.permission;

import net.blueshell.common.identity.IdentityProvider;
import net.blueshell.db.BaseModel;
import net.blueshell.db.BaseModelService;
import org.springframework.core.GenericTypeResolver;
import org.springframework.security.core.Authentication;

import java.util.HashMap;
import java.util.Map;
import java.util.function.BiFunction;

public abstract class BasePermissionEvaluator<T extends BaseModel<ID>, ID, S extends BaseModelService<T, ID, ?>> extends IdentityProvider {

    protected final S service;
    protected final Map<String, BiFunction<T, Authentication, Boolean>> permissionsMap = new HashMap<>() {
    };
    public final Class<T> domainType;

    public BasePermissionEvaluator(S service) {
        this.service = service;
        this.domainType = determineDomainType();
    }

    private Class<T> determineDomainType() {
        Class<?>[] resolvedTypes = GenericTypeResolver.resolveTypeArguments(getClass(), BasePermissionEvaluator.class);
        if (resolvedTypes == null || resolvedTypes.length < 1) {
            throw new IllegalStateException("Unable to determine domain type for " + getClass().getName());
        }
        @SuppressWarnings("unchecked")
        Class<T> castedType = (Class<T>) resolvedTypes[0];
        return castedType;
    }

    boolean supports(Class<?> domainClass) {
        return domainType.isAssignableFrom(domainClass);
    }

    public abstract boolean hasPermission(Authentication authentication, Object targetDomainObject, String string);
    public abstract boolean hasPermissionId(Authentication authentication, Object targetId, String string);
}

File: api/src/main/kotlin/net/blueshell/api/platform/config/permission/BasePermissionEvaluator.kt
package net.blueshell.api.platform.config.permission

import net.blueshell.api.feature.auth.security.IdentityProvider
import net.blueshell.api.shared.model.Identifiable
import net.blueshell.api.shared.service.BaseModelService
import org.springframework.core.GenericTypeResolver
import org.springframework.security.core.Authentication

abstract class BasePermissionEvaluator<T : Identifiable<ID>, ID, S : BaseModelService<T, ID, *>>(protected val service: S) :
    IdentityProvider() {
    val domainType: Class<T>

    init {
        this.domainType = determineDomainType()
    }

    private fun determineDomainType(): Class<T> {
        val resolvedTypes = GenericTypeResolver.resolveTypeArguments(javaClass, BasePermissionEvaluator::class.java)
        check(!(resolvedTypes == null || resolvedTypes.size < 1)) { "Unable to determine domain type for " + javaClass.name }
        return resolvedTypes[0] as Class<T>
    }

    fun supports(domainClass: Class<*>): Boolean {
        return domainType.isAssignableFrom(domainClass)
    }

    abstract fun hasPermission(authentication: Authentication?, targetDomainObject: Any?, string: String?): Boolean

    abstract fun hasPermissionId(authentication: Authentication?, targetId: Any?, string: String?): Boolean
}